syntax = "proto3";

package infinitepickaxe;

import "google/protobuf/wrappers.proto";

// 공통 Envelope: msg_type으로 메시지 구분
message Envelope {
  string msg_type = 1;  // HANDSHAKE, HEARTBEAT, MINE_START, ...
  uint32 version = 2;   // 프로토콜 버전
  uint32 seq = 3;       // 시퀀스 번호 (옵션, 0이면 미사용)
  uint64 timestamp = 4; // 클라이언트 Unix timestamp(초)
  bytes payload = 5;    // 직렬화된 메시지 페이로드
}

// -------------- 접속 / 세션 --------------

message HandshakeReq {
  string jwt = 1;
  string client_version = 2;
  string device_id = 3;
}

message PickaxeSlotInfo {
  uint32 slot_index = 1;           // 1-4
  uint32 level = 2;                // 곡괭이 레벨
  uint64 attack_power = 3;         // 공격력
  uint32 attack_speed_x100 = 4;    // 공격속도 * 100 (예: 1.5 APS = 150)
  uint64 dps = 5;                  // 이 슬롯의 DPS = (attack_power * attack_speed_x100) / 100
  bool is_unlocked = 6;            // 슬롯 해금 여부
}

message UserDataSnapshot {
  google.protobuf.UInt64Value gold = 1;
  google.protobuf.UInt32Value crystal = 2;
  repeated bool unlocked_slots = 3;              // [true, true, false, false] 형태
  google.protobuf.UInt32Value current_mineral_id = 4;
  google.protobuf.UInt64Value mineral_hp = 5;
  google.protobuf.UInt64Value mineral_max_hp = 6;

  // 슬롯 정보 (4개)
  repeated PickaxeSlotInfo pickaxe_slots = 7;    // 최대 4개
  uint64 total_dps = 8;                          // 모든 슬롯의 DPS 합계

  google.protobuf.UInt64Value server_time = 9;

  // 미션 관련
  uint32 ads_watched_today = 10;
  uint32 mission_rerolls_used = 11;
  uint32 offline_bonus_hours = 12;
}

message HandshakeRes {
  bool ok = 1;
  string error = 2;        // INVALID_JWT, BANNED, EXPIRED ...
  UserDataSnapshot snapshot = 3;
}

message Heartbeat {
  uint64 client_time_ms = 1;
}

message HeartbeatAck {
  uint64 server_time_ms = 1;
}

// -------------- 광물 정보 --------------

message MineralListRequest {
  // 빈 요청
}

message MineralInfo {
  uint32 mineral_id = 1;
  string name = 2;              // "약한 돌", "돌", "석탄" 등
  uint64 max_hp = 3;
  uint64 gold_reward = 4;
  string icon_name = 5;         // 아이콘 리소스 이름 (옵션)
}

message MineralListResponse {
  repeated MineralInfo minerals = 1;
}

// 광물 변경 (MineralSelectModal에서 "선택" 버튼)
message ChangeMineralRequest {
  uint32 mineral_id = 1;
}

message ChangeMineralResponse {
  bool success = 1;
  uint32 mineral_id = 2;
  uint64 mineral_hp = 3;        // 새 광물의 현재 HP
  uint64 mineral_max_hp = 4;    // 새 광물의 최대 HP
  string error_code = 5;        // "INVALID_MINERAL" 등
}

// -------------- 채굴 --------------

message MiningStart {
  uint32 mineral_id = 1;        // 채굴할 광물 (모든 슬롯이 동일 광물 채굴)
}

message MiningSync {
  uint32 mineral_id = 1;
  uint64 client_hp = 2;
  uint64 client_timestamp = 3;
}

message MiningUpdate {
  uint32 mineral_id = 1;
  uint64 current_hp = 2;
  uint64 max_hp = 3;
  uint64 damage_dealt = 4;      // 이번 틱의 총 데미지
  uint64 total_dps = 5;         // 모든 슬롯의 DPS 합계
  uint64 server_timestamp = 6;
}

message MiningComplete {
  uint32 mineral_id = 1;
  uint64 gold_earned = 2;
  uint64 total_gold = 3;
  uint64 mining_count = 4;
  uint32 respawn_time = 5;
  uint64 server_timestamp = 6;
}

// -------------- 슬롯 정보 조회 --------------

// 모든 슬롯 정보 조회 (한번에)
message AllSlotsRequest {
  // 빈 요청
}

message AllSlotsResponse {
  repeated PickaxeSlotInfo slots = 1;  // 최대 4개
  uint64 total_dps = 2;                // 모든 슬롯의 DPS 합계
}

// -------------- 강화 --------------

message UpgradePickaxe {
  uint32 slot_index = 1;       // 강화할 슬롯 (1-4)
  uint32 target_level = 2;     // 목표 레벨
}

message UpgradeResult {
  bool success = 1;
  uint32 slot_index = 2;
  uint32 new_level = 3;
  uint64 new_attack_power = 4;        // 강화 후 공격력
  uint32 new_attack_speed_x100 = 5;   // 강화 후 공격속도 * 100
  uint64 new_dps = 6;                 // 강화 후 이 슬롯의 DPS
  uint64 new_total_dps = 7;           // 강화 후 모든 슬롯의 총 DPS
  uint64 gold_spent = 8;
  uint64 remaining_gold = 9;
  string error_code = 10;
  uint32 base_rate_bp = 11;           // 기본 확률 (basis point 10000)
  uint32 bonus_rate_bp = 12;          // 보너스 확률
  uint32 final_rate_bp = 13;          // 최종 확률
  uint32 pity_bonus = 14;             // 천장 보너스
}

// -------------- 미션 --------------

message MissionClaim {
  uint32 mission_index = 1;
}

message MissionReroll {
  bool use_ad = 1;
}

message MissionEntry {
  uint32 index = 1;
  string type = 2;
  string description = 3;
  uint32 target = 4;
  uint32 current = 5;
  uint32 reward_crystal = 6;
  bool completed = 7;
  bool claimed = 8;
}

message MissionUpdate {
  repeated MissionEntry missions = 1;
  bool milestone_completed_3 = 2;
  bool milestone_completed_5 = 3;
  bool milestone_completed_7 = 4;
  uint32 offline_bonus_hours = 5;
  uint64 reset_time = 6; // unix seconds
}

// 마일스톤 보상 수령 (QuestTab)
message MilestoneClaim {
  uint32 milestone_count = 1;  // 3, 5, 7
}

message MilestoneClaimResult {
  bool success = 1;
  uint32 milestone_count = 2;
  uint32 offline_hours_gained = 3;  // 오프라인 보상 시간 증가량
  uint32 total_offline_hours = 4;   // 현재 총 오프라인 보상 시간
  string error_code = 5;
}

// -------------- 슬롯 해금 --------------

message SlotUnlock {
  uint32 slot_index = 1;
}

message SlotUnlockResult {
  bool success = 1;
  uint32 slot_index = 2;
  uint32 crystal_spent = 3;
  uint32 remaining_crystal = 4;
  string error_code = 5;
}

// -------------- 광고 시청 --------------

message AdWatchComplete {
  uint32 ad_tier = 1;  // 1, 2, 3
}

message AdWatchResult {
  bool success = 1;
  uint32 ad_tier = 2;
  uint32 crystal_earned = 3;
  uint32 total_crystal = 4;
  uint32 ads_watched_today = 5;
  string error_code = 6;
}

// -------------- 실시간 재화 동기화 --------------

message CurrencyUpdate {
  google.protobuf.UInt64Value gold = 1;
  google.protobuf.UInt32Value crystal = 2;
  string reason = 3;  // "mining", "upgrade", "ad", "mission", "milestone" 등
}

// -------------- 오프라인 보상 --------------

message OfflineRewardRequest {
  string request_type = 1; // "calculate" 등
}

message OfflineReward {
  uint32 offline_seconds = 1;
  uint64 gold_earned = 2;
  uint32 mining_cycles = 3;
  uint32 mineral_id = 4;
  double efficiency = 5;
  uint64 new_total_gold = 6;
}

// -------------- 공통 에러 --------------

message Error {
  string error_code = 1;    // AUTH_FAILED, INVALID_PACKET, INSUFFICIENT_GOLD 등
  string error_message = 2; // 상세 메시지
  bytes detail = 3;         // 선택적 추가 JSON/PROTO
}
